routing/missioncontrol_store_test.go

```diff
@@ -148,3 +150,242 @@ func TestMissionControlStore(t *testing.T) {
 			spew.Sdump(results[1]))
 	}
 }
+
+// TestMissionControlStoreFlushing asserts the periodic flushing of the store
+// works correctly.
+func TestMissionControlStoreFlushing(t *testing.T) {
+	// Set time zone explicitly to keep test deterministic.
+	time.Local = time.UTC
+
+	file, err := os.CreateTemp("", "*.db")
+	require.NoError(t, err)
+
+	dbPath := file.Name()
+	t.Cleanup(func() {
+		require.NoError(t, file.Close())
+		require.NoError(t, os.Remove(dbPath))
+	})
+
+	db, err := kvdb.Create(
+		kvdb.BoltBackendName, dbPath, true, kvdb.DefaultDBTimeout,
+	)
+	require.NoError(t, err)
+	t.Cleanup(func() {
+		require.NoError(t, db.Close())
+	})
+
+	const flushInterval = time.Second
+
+	store, err := newMissionControlStore(db, testMaxRecords, flushInterval)
+	require.NoError(t, err)
+
+	testRoute := route.Route{
+		SourcePubKey: route.Vertex{1},
+		Hops: []*route.Hop{
+			{
+				PubKeyBytes:   route.Vertex{2},
+				LegacyPayload: true,
+			},
+		},
+	}
+
+	failureSourceIdx := 1
+	failureDetails := lnwire.NewFailIncorrectDetails(100, 1000)
+	var lastID uint64 = 0
+	nextResult := func() *paymentResult {
+		lastID += 1
+		return &paymentResult{
+			route:            &testRoute,
+			failure:          failureDetails,
+			failureSourceIdx: &failureSourceIdx,
+			id:               lastID,
+			timeReply:        testTime,
+			timeFwd:          testTime.Add(-time.Minute),
+		}
+	}
+
+	// Helper to assert the number of results is correct.
+	assertResults := func(wantCount int) {
+		t.Helper()
+		err := wait.NoError(func() error {
+			results, err := store.fetchAll()
+			if err != nil {
+				return err
+			}
+			if wantCount != len(results) {
+				return fmt.Errorf("wrong nb of results: got "+
+					"%d, want %d", len(results), wantCount)
+			}
+			if len(results) == 0 {
+				return nil
+			}
+			gotLastID := results[len(results)-1].id
+			if len(results) > 0 && gotLastID != lastID {
+				return fmt.Errorf("wrong id for last item: "+
+					"got %d, want %d", gotLastID, lastID)
+			}
+
+			return nil
+		}, flushInterval*5)
+		require.NoError(t, err)
+	}
+
+	// Run the store.
+	store.run()
+	time.Sleep(flushInterval)
+
+	// Wait for the flush interval. There should be no records.
+	assertResults(0)
+
+	// Store a result and check immediately. There still shouldn't be
+	// any results stored (flush interval has not elapsed).
+	store.AddResult(nextResult())
+	assertResults(0)
+
+	// Assert that eventually the result is stored after being flushed.
+	assertResults(1)
+
+	// Store enough results that fill the max nb of results.
+	for i := 0; i < testMaxRecords+1; i++ {
+		store.AddResult(nextResult())
+	}
+	assertResults(testMaxRecords)
+
+	// Finally, stop the store to recreate it.
+	store.stop()
+
+	// Recreate store.
+	store, err = newMissionControlStore(db, testMaxRecords, flushInterval)
+	require.NoError(t, err)
+	store.run()
+	defer store.stop()
+	time.Sleep(flushInterval)
```

Comment:
For the case where no updates happen, there is no way to know a flush happened. This is here to assert that, even after the flush interval elapses, no results are magically appearing.

The "correct" way to assert this would be to add a hook to be called after every flush, but that would only be useful for this specific test and would add an unnecessary (albeit very very small) overhead on production. And in any case, we don't even expect any results for the baseline: something would have to be very very wrong for there to appear a result here. So I took a shortcut.

Reply:
(deleted the prior comment because I was looking at the wrong code)

Reply:
This is here to assert that even after the first flush interval (i.e. the first potential flush), the results are correct (i.e. results have not been added or removed) after recreation.

The "correct" way to go about this would be to add a hook called after after every flush to determine when a flush happened and then perform the check. This would only be useful during testing and would increase complexity of the MC structure and add a (very very small) overhead on production, so I chose to take a shortcut here, since we're just asserting the baseline (that the results are the same).

Note that this (adding the hook and waiting on it) would mean the assertion would change during the PR: before my PR, the flush is continuous (so we'd have to assert the flush happened) and afterwards, the flush is only after a change (so we'd have to assert that no flush happened) - meaning the test would get more coupled to the implementation, also making it ever slightly more brittle.

If you feel this is the right approach here though, I'll make the appropriate changes.

Reply:
ok cool - makes sense - fine to leave as is I think! Just wanted to understand the reason 👍